File: pb/toutiao_dmp_test.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
import time
import base64
import chardet
import toutiao_dmp_pb2  # 由pb文件生成的python代码, 使用Protocol Buffer2


def generate_valid_file():
    dmp_data = toutiao_dmp_pb2.DmpData()

    id_item1 = dmp_data.idList.add()
    id_item1.dataType = toutiao_dmp_pb2.IdItem.IMEI
    id_item1.id = '356145080566857'
    id_item1.tags.append("1")
    id_item1.tags.append("2")
    id_item1.timestamp = int(time.time())

    id_item2 = dmp_data.idList.add()
    id_item2.dataType = toutiao_dmp_pb2.IdItem.IDFA
    id_item2.id = '1E2DFA89-496A-47FD-9941-DF1FC4E6484A'
    id_item2.tags.append("3")
    id_item2.tags.append("4")

    binary_string = dmp_data.SerializeToString()
    result_string = base64.b64encode(binary_string)

    dmp_data2 = toutiao_dmp_pb2.DmpData()

    id_item21 = dmp_data2.idList.add()
    id_item21.dataType = toutiao_dmp_pb2.IdItem.IMEI
    id_item21.id = '136145080566857'
    id_item21.tags.append("5")
    id_item21.tags.append("6")
    id_item21.timestamp = int(time.time())

    id_item22 = dmp_data2.idList.add()
    id_item22.dataType = toutiao_dmp_pb2.IdItem.IDFA
    id_item22.id = '642DFA89-496A-47FD-9941-DF1FC4E6484A'
    id_item22.tags.append("7")
    id_item22.tags.append("8")

    binary_string2 = dmp_data2.SerializeToString()
    print()

    result_string2 = base64.b64encode(binary_string2)

    target_file = open('./target_pb2', 'w')
    target_file.write(str(result_string, encoding="utf-8"))
    target_file.write('\n')
    target_file.write(str(result_string2, encoding="utf-8"))
    target_file.write('\n')
    target_file.close()


if __name__ == '__main__':
    generate_valid_file()

File: pb/toutiao_dmp_test.go
package pb

import (
	"encoding/base64"
	"github.com/golang/protobuf/proto"
	"log"
	"os"
	"testing"
	"time"
)

func TestToutiaoDMPGenerate(t *testing.T) {
	dmpData1 := DmpData{}
	dmpData2 := DmpData{}

	var (
		id1 = "356145080566857"
		ts1 = uint32(time.Now().Unix())
		dt1 = IdItem_IMEI
	)

	var (
		id2 = "1E2DFA89-496A-47FD-9941-DF1FC4E6484A"
		ts2 = uint32(time.Now().Unix())
		dt2 = IdItem_IDFA
	)

	var (
		id3 = "136145080566857"
		ts3 = uint32(time.Now().Unix())
		dt3 = IdItem_IMEI
	)

	var (
		id4 = "642DFA89-496A-47FD-9941-DF1FC4E6484A"
		ts4 = uint32(time.Now().Unix())
		dt4 = IdItem_IDFA
	)

	idItem1 := IdItem{
		Timestamp: &ts1,
		DataType:  &dt1,
		Id:        &id1,
		Tags:      []string{"信用卡", "理财"},
	}

	idItem2 := IdItem{
		Timestamp: &ts2,
		DataType:  &dt2,
		Id:        &id2,
		Tags:      []string{"黄金", "理财"},
	}

	idItem3 := IdItem{
		Timestamp: &ts3,
		DataType:  &dt3,
		Id:        &id3,
		Tags:      []string{"信用卡", "股票"},
	}

	idItem4 := IdItem{
		Timestamp: &ts4,
		DataType:  &dt4,
		Id:        &id4,
		Tags:      []string{"黄金", "理财"},
	}

	dmpData1.IdList = append(dmpData1.IdList, &idItem1)
	dmpData1.IdList = append(dmpData1.IdList, &idItem2)

	dmpData2.IdList = append(dmpData2.IdList, &idItem3)
	dmpData2.IdList = append(dmpData2.IdList, &idItem4)

	binaryString1, _ := proto.Marshal(&dmpData1)
	resultString1 := base64.StdEncoding.EncodeToString(binaryString1)

	binaryString2, _ := proto.Marshal(&dmpData2)
	resultString2 := base64.StdEncoding.EncodeToString(binaryString2)

	file, err := os.OpenFile("target_pb2", os.O_CREATE|os.O_WRONLY, 0666)
	if err != nil {
		log.Panic(err)
	}

	defer file.Close()

	if _, err := file.WriteString(resultString1 + "\n"); err != nil {
		log.Panic(err)
	}
	//if _, err := file.WriteString("\n"); err != nil {
	//	log.Panic(err)
	//}
	if _, err := file.WriteString(resultString2 + "\n"); err != nil {
		log.Panic(err)
	}
	//if _, err := file.WriteString("\n"); err != nil {
	//	log.Panic(err)
	//}
}
